betty/plugin/__init__.py

```diff
@@ -0,0 +1,63 @@
+"""
+The Plugin API.
+
+Plugins allow third-party code (e.g. your own Python package) to add functionality
+to Betty.
+"""
+
+from collections.abc import Iterable
+from typing import TypeVar, Generic
+
+from betty.locale import Localizable
+
+
+class Plugin:
+    """
+    A plugin.
+    """
+
+    @classmethod
+    def plugin_id(cls) -> str:
+        """
+        Get the plugin ID.
+
+        IDs are unique per plugin type:
+        - A plugin repository MUST at most have a single plugin for any ID.
+        - Different plugin repositories MAY each have a plugin with the same ID.
+        """
+        raise NotImplementedError(repr(cls))
+
+    @classmethod
+    def plugin_label(cls) -> Localizable:
+        """
+        Get the human-readable short plugin label.
+        """
+        raise NotImplementedError(repr(cls))
+
+    @classmethod
+    def plugin_description(cls) -> Localizable | None:
+        """
+        Get the human-readable long plugin description.
+        """
+        return None
+
+
+PluginT = TypeVar("PluginT", bound=Plugin)
+
+
+class PluginRepository(Generic[PluginT]):
+    """
+    Discover and manage plugins.
+    """
+
+    async def get(self, plugin_id: str) -> type[PluginT]:
```

Comment:
Add an exception that MUST be raised when an ID is given for which no plugin exists.